Python/ceval.c

```diff
@@ -2207,6 +2208,34 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, _PyInterpreterFrame *frame, int
             NOTRACE_DISPATCH();
         }
 
+        TARGET(BINARY_SUBSCR_ASCII_INT) {
+            assert(cframe.use_tracing == 0);
+            PyObject *sub = TOP();
+            PyObject *unicode = SECOND();
+            DEOPT_IF(!PyLong_CheckExact(sub), BINARY_SUBSCR);
+            DEOPT_IF(!PyUnicode_CheckExact(unicode), BINARY_SUBSCR);
+            DEOPT_IF(!PyUnicode_IS_ASCII(unicode), BINARY_SUBSCR);
+
+            // Deopt unless 0 <= sub < ascii->length
+            PyASCIIObject *ascii = _PyASCIIObject_CAST(unicode);
+            Py_ssize_t signed_magnitude = Py_SIZE(sub);
+            DEOPT_IF(((size_t)signed_magnitude) > 1, BINARY_SUBSCR);
+            assert(((PyLongObject *)_PyLong_GetZero())->ob_digit[0] == 0);
+            Py_ssize_t index = ((PyLongObject*)sub)->ob_digit[0];
+            DEOPT_IF(index >= ascii->length, BINARY_SUBSCR);
+            STAT_INC(BINARY_SUBSCR, hit);
+            Py_UCS4 ch = PyUnicode_1BYTE_DATA(ascii)[index];
```

Comment:
This still has branches as it checks if the string is ascii or not. The following should be much faster.
```suggestion
            const void *data = ((void*)(ascii + 1));
            Py_UCS1 ch = ((Py_UCS1*)data)[index];
```